File: src/previous/day3/Stack.ts
type Node<T> = {
    value: T
    prev?: Node<T>
}

export default class Stack<T> {
    public length: number
    public head?: Node<T>
    public tail?: Node<T>
    constructor() {
        this.length = 0
        this.head = this.tail = undefined
    }

    // 2 >> 3 >> 5 << push in
    push(item: T): void {
        const node = { value: item } as Node<T>
        this.length++
        if (!this.tail) {
            this.head = this.tail = node
            return
        }
        node.prev = this.tail
        this.tail = node
    }

    // 2 >> 3 >> 5 << pop the last
    // 2 << pop the last
    pop(): T | undefined {
        if (!this.tail) {
            return undefined
        }
        this.length--
        const curr = this.tail
        if (!this.tail.prev) {
            this.head = this.tail = undefined
            return curr.value
        }

        this.tail = this.tail.prev
        return curr.value
    }
    peek(): T | undefined {
        return this.head?.value
    }
}

File: go/day02/Stack/Stack.go
package stack

type Node[T comparable] struct {
	prev  *Node[T]
	value T
}

type Stack[T comparable] struct {
	head   *Node[T]
	tail   *Node[T]
	length int
}

func MakeStack[T comparable]() *Stack[T] {
	return &Stack[T]{
		length: 0,
	}
}

func (s *Stack[T]) Push(item T) {
	node := &Node[T]{
		value: item,
	}
	s.length++

	if s.tail == nil {
		s.head = node
		s.tail = node
	}

	node.prev = s.tail
	s.tail = node
}

func (s *Stack[T]) Pop() T {
	var out T
	if s.tail == nil {
		return out
	}

	out = s.tail.value
	s.length--
	if s.length == 0 {
		s.head = nil
		s.tail = nil
		return out
	}

	s.tail = s.tail.prev
	return out
}

func (s *Stack[T]) Peek() T {
	var out T
	if s.head == nil {
		return out
	}

	out = s.head.value
	return out
}
